### PR TITLE
feat: support subjectDirectoryAttributes and associatedInformation extensions

### LINKED ISSUE #3473: Adding Support for x509v3 Extension: Subject Directory Attributes
Hi there,

please integrate the needed functions, structures and data types for full handling (read / parse and create) the X509v3 Subject Directory Attributes Extensions within OpenSSL. 

Or maybe someone of the great community can do this and make a pull request? I can't do it by own, because i don't have enough skills for that.

### PR DESCRIPTION
**Pull request by JonathanWilbur**
Fixes #3473 

This PR adds support for the `subjectDirectoryAttributes` and `associatedInformation` X.509v3 extensions. These extensions allow an X.509 public key certificate to include arbitrary directory attributes: the former associated with the subject, and the latter containing arbitrary information that does not necessarily describe the subject.

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by JonathanWilbur** [comment]
Per the failing `check_docs` job, it says that `OSSL_print_attribute_value` is not public, even though I added it to `include/crypto/x509.h`. Do I need to do something else to get it recognized as public?

**Comment by mattcaswell** [comment]
crypto/x509.h is not a public header file. Perhaps you meant to add it to openssl/x509.h? Add it to the correct header. Then run `make update` and commit any updates made.

File: doc/man3/OSSL_print_attribute_value.pod
Comment (t8m) [inline doc/man3/OSSL_print_attribute_value.pod:7]: Hmm, do you really intend to make it a public function? In that case it must be placed into openssl/x509.h and not crypto/x509.h

However I am not sure it should be.

**Review by t8m — CHANGES_REQUESTED** [review]

File: doc/man3/OSSL_print_attribute_value.pod
Comment (FdaSilvaYY) [inline doc/man3/OSSL_print_attribute_value.pod:6]: X.509, no ?

Comment (JonathanWilbur) [reply doc/man3/OSSL_print_attribute_value.pod:6]: I said X.500 because X.509 is a part of a larger X.500-series of specifications in which attribute types and values are introduced. Technically, the attributes used in these extensions are actually introduced in X.501. X.509 is just for cryptography / PKI / security aspects of the series.

Comment (JonathanWilbur) [reply doc/man3/OSSL_print_attribute_value.pod:6]: That said, I am sure others would have similar confusion, so I'd be happy to change this if you still insist.

File: crypto/x509/x_attrib.c
Comment (JonathanWilbur) [reply crypto/x509/x_attrib.c:264]: I actually wanted to ask about this. I figured this would be a contentious matter.

1. OpenSSL doesn't have any concept of `DATE`, `TIME-OF-DAY`, `DATE-TIME`, `DURATION`, `OID-IRI` and `RELATIVE-OID-IRI`. These were added as universal types in newer versions of ASN.1. Those types all represented as some sort of string, so I would think if OpenSSL supports those eventually, it will represent them as strings as well. (I am willing to make this PR, by the way. That should be simple enough, assuming we don't care to do any validation on them!). AFAIK, these are not currently used in any extensions or attributes published to date.
2. `UniversalString` and `BMPString` are capable of representing Unicode characters. How does OpenSSL perform output? Can I assume the terminal takes UTF-8 output? It seems like `UTF8String`s are written to `stdout` elsewhere in OpenSSL.
3. `CHARACTER STRING`, `EXTERNAL`, and `EMBEDDED PDV` are constructed types that are rarely used at all. I was thinking that, if I produced an ASN.1 dump for these, it would not be obvious how to differentiate these from `SET` or `SEQUENCE`, but then I realized that this is also true between `SET` and `SEQUENCE`. Should I prefix the ASN.1 dumps with something like `SET:` or `SEQUENCE:`?

File: doc/man3/OSSL_print_attribute_value.pod
Comment (JonathanWilbur) [reply doc/man3/OSSL_print_attribute_value.pod:7]: I am open to it being a private function, by the way. It just seemed kind of foundational to be able to print attribute values. What makes it preferable to be public or private? Is there some guiding principle for this?

File: crypto/x509/x_attrib.c
Comment (t8m) [reply crypto/x509/x_attrib.c:271]: > Can I assume the terminal takes UTF-8 output?

Yes, that is what other places in OpenSSL assume, so IMO it can be assumed here as well.

File: doc/man3/OSSL_print_attribute_value.pod
Comment (t8m) [reply doc/man3/OSSL_print_attribute_value.pod:7]: If there is a clear use case for 3rd party applications and the API is not controversial/unfinished by other means, it can be public.

Otherwise if you need something internally, by default an internal API should be added.

Comment (JonathanWilbur) [reply doc/man3/OSSL_print_attribute_value.pod:7]: Okay. Yeah, I do suspect this to be a controversial API. I imagine everybody has a different opinion on how these values should be printed. And I suspect this function will have a long tail of small tweaks throughout time. I'll make it private, since it matches both criteria very well. Thank you!

File: doc/man3/OSSL_print_attribute_value.pod
Comment (JonathanWilbur) [reply doc/man3/OSSL_print_attribute_value.pod:6]: I will interpret your thumbs-up as assent and mark this issue resolved. It might not matter anymore, since I am going to make `OSSL_print_attribute_value` a private function now and remove this documentation.

File: doc/man3/OSSL_print_attribute_value.pod
Comment (JonathanWilbur) [reply doc/man3/OSSL_print_attribute_value.pod]: Removed from the public API.

**Comment by t8m** [comment]
Do not remove the documentation completely, just move it into doc/internal

File: include/crypto/x509.h
Comment (t8m) [inline include/crypto/x509.h:392]: Please add `ossl_` prefix (lowercase).

**Review by t8m — CHANGES_REQUESTED** [review]

File: doc/internal/man3/ossl_print_attribute_value.pod
Comment (t8m) [inline doc/internal/man3/ossl_print_attribute_value.pod:16]:
```suggestion
ossl_print_attribute_value() prints an X.500 directory value, which is an
```

File: doc/internal/man3/ossl_print_attribute_value.pod
Comment (t8m) [inline doc/internal/man3/ossl_print_attribute_value.pod:30]:
```suggestion
The I<out> argument is a B<BIO> pointer for printing the output. The I<obj_nid>
argument is the NID of the attribute type object identifier. The ASN.1 value
itself is passed in I<av> and the level of desired indentation in terms of the
number of spaces is specified in I<indent>.
```

File: doc/internal/man3/ossl_print_attribute_value.pod
Comment (t8m) [inline doc/internal/man3/ossl_print_attribute_value.pod:44]: `ossl_...` or you can just remove the HISTORY section as it is not relevant for internal documentation.

File: include/crypto/x509.h
Comment (t8m) [inline include/crypto/x509.h:392]: Nit: please split the long line.

File: crypto/x509/x_attrib.c
Comment (t8m) [inline crypto/x509/x_attrib.c:106]: Nit: please split the long line

File: crypto/x509/x_attrib.c
Comment (t8m) [reply crypto/x509/x_attrib.c:271]: Anyway I am OK with postponing this.

**Review by t8m — CHANGES_REQUESTED** [review]
A few last touches

**Comment by t8m** [comment]
I'd also suggest squashing the commits before applying the last fixup (I recommend to use `git commit --fixup` for that.)

File: crypto/x509/x_attrib.c
Comment (JonathanWilbur) [reply crypto/x509/x_attrib.c:274]: Yeah, I looked into implementing UniversalString and BMPString conversion to UTF-8, and I think that is a little risky, so I'd like to do that in a separate PR so it gets a lot more scrutiny.  And the other types aren't implemented in OpenSSL yet, which I would also like to do in a separate PR. I'll mark this as resolved, then.

**Comment by JonathanWilbur** [comment]
I forgot to `--fixup` when I committed and pushed. Do you want me to squash the commit history?

**Comment by t8m** [comment]
> I forgot to `--fixup` when I committed and pushed. Do you want me to squash the commit history?

no problem

**Review by t8m — APPROVED** [review]

**Comment by JonathanWilbur** [comment]
The test failure appears to have nothing to do with my code. @nhorman would you care to give this one a review?

**Comment by nhorman** [comment]
Its been happening all day, Canonicals repo had a hiccup. I'm restarting

File: crypto/x509/x_attrib.c
Comment (nhorman) [inline crypto/x509/x_attrib.c:171]: This series of cases is a bit confusing to me.  You have a case statement here that identifies the type of value that a given ASN1 string is, but in the event it can't be decoded, we just print the string.  Shouldn't that be an error condition?  Or are there cases in which this is normal?

File: crypto/x509/x_attrib.c
Comment (nhorman) [inline crypto/x509/x_attrib.c]: shouldn't there be a recursive call here to print the contents of the object?

File: crypto/x509/x_attrib.c
Comment (nhorman) [inline crypto/x509/x_attrib.c]: BMP String?  Isn't that potentially....large?  Nothing wrong with printing it if its properly converted to ascii of course, but is there value in printing what I am assuming to be a bitmap as a character string?

**Review by nhorman — COMMENTED** [review]
This is going to take some time to get through, so I'll start with these comments and see what you think @JonathanWilbur

File: crypto/x509/v3_sda.c
Comment (JonathanWilbur) [reply crypto/x509/v3_sda.c:65]: We could not print. I felt like it was better to be explicit rather than to leave the user thinking they had an empty value, such as an empty string or zero-length octet string. But I am open to removing this if you'd prefer that.

File: crypto/x509/x_attrib.c
Comment (JonathanWilbur) [reply crypto/x509/x_attrib.c]: You're absolutely right. I will fix that.

File: crypto/x509/v3_sda.c
Comment (nhorman) [reply crypto/x509/v3_sda.c:65]: No, I'm leaving that up to you, I was just asking the question

File: crypto/x509/x_attrib.c
Comment (JonathanWilbur) [reply crypto/x509/x_attrib.c:171]: I feel like you are looking at an old version of the file here. I had a mistake in here earlier, which I fixed. (`str` was never set.) What my current code does (and what my original intent was) is print the `INTEGER` as a decimal if it can fit in `int64_t`, but in hexadecimal otherwise. I will mark this as resolved unless you look at my newest code and believe it to still be a problem.

File: crypto/x509/x_attrib.c
Comment (JonathanWilbur) [reply crypto/x509/x_attrib.c]: This also appears to be an outdated view of my code. I removed this.

File: crypto/x509/x_attrib.c
Comment (JonathanWilbur) [reply crypto/x509/x_attrib.c]: I don't think so. `V_ASN1_OBJECT` is an object identifier, not another recursively structured type. This function also prints attribute values, not ASN.1 values, so recursion doesn't really make sense here unless a particular attribute value is composed of further attributes.

It does however, print structured types in the `V_ASN1_SEQUENCE` and `V_ASN1_SET` cases using the same dump that `asn1dump` uses. But again, recursion doesn't really make sense here, however, because these structure types are not necessarily (and are almost never, if not never) attributes themselves.

I will make this as resolved, but feel free to un-resolve if you have more questions or issues with this.

**Comment by JonathanWilbur** [comment]
All feedback addressed and ready for review.

**Review by t8m — APPROVED** [review]

File: crypto/x509/x_attrib.c
Comment (nhorman) [inline crypto/x509/x_attrib.c:69]: This is minor, but since we have OPENSSL_buf2hexstr_ex already available, it may be worth leveraging that here.  Its minor though, may not be worth doing.

**Review by nhorman — APPROVED** [review]

File: crypto/x509/x_attrib.c
Comment (FdaSilvaYY) [inline crypto/x509/x_attrib.c:139]: Why don't you send directly 'value' ptr here ?
It will spare you `av->value.sequence->data` pointer restoration a few lines below.

File: crypto/x509/standard_exts.h
Comment (FdaSilvaYY) [inline crypto/x509/standard_exts.h:85]: last comma is useless

File: crypto/x509/x_attrib.c
Comment (FdaSilvaYY) [inline crypto/x509/x_attrib.c:76]: Start with `int result = -1;` ...

File: crypto/x509/x_attrib.c
Comment (FdaSilvaYY) [inline crypto/x509/x_attrib.c:86]: set `result = 0;` just after `OPENSSL_free()`.
So you can remove the `else` block.

File: crypto/x509/x_attrib.c
Comment (FdaSilvaYY) [inline crypto/x509/x_attrib.c:94]: `const ASN1_OBJECT *oid` ?

**Review by FdaSilvaYY — COMMENTED** [review]
Good job, just a few nits.

File: crypto/x509/standard_exts.h
Comment (JonathanWilbur) [reply crypto/x509/standard_exts.h:85]: This is just a personal preference, to some extent, but the reason I prefer the trailing comma is that it makes version control diffs smaller and simpler: https://eslint.org/docs/latest/rules/comma-dangle.

**Comment by JonathanWilbur** [comment]
Actually, I am sorry to do this to the reviewers, but I would like more time to work on this before it is merged in. On a second look, I don't feel like I did a thorough job on this. Namely, what I would like to change is:

1. Printing a `BIT STRING` using bits rather than as hex.
2. Something about `T61String`. I just realized that the higher 128 characters of a T.61 are not valid UTF-8, so attempting to print this while assuming the terminal output takes UTF-8 is going to produce UTF-8 error characters (or however the terminal handles it). On the other hand, there's nothing stopping certificates with invalid UTF-8 in a `UTF8String` from being printed either. Not sure what to do here, since a hex string could interpreted as the actual string value. If I wrap it in quotes, I think that creates a standard that I'll have to do for all other strings, and I might have to deal with quote escaping.
3. Handle non-`UNIVERSAL` tags. This is an obvious one I missed.
4. Incorporate @FdaSilvaYY 's and @nhorman 's feedback.
5. More testing. I noticed that my test certificates do not test small or large integer, bit strings, or non-DN `SET` or `SEQUENCE`.

I would also be happy to address these things in a quick follow-up PR if you guys would rather get this in sooner, since it was already approved.

**Comment by JonathanWilbur** [comment]
Here is what the current output looks like, just for your visual review, by the way:

```
            X509v3 Subject Directory Attributes: 
                commonName:
                    Steve Brule
                    Dr. Steve Brule
                    Dr. Steve Brule, from Brules Rules
                    The givenName attribute below is intentionally empty
                2.5.4.106:
                    2.5.4.3 (commonName)
                owner:
                    CN=Hi mom
                givenName:
                    <No Values>
                localityName:
                    Funkytown
```

**Comment by nhorman** [comment]
It's got 2 reviews, id submit a follow up PR

**Comment by JonathanWilbur** [comment]
Okay, sounds good. Thank you!

**Comment by JonathanWilbur** [comment]
@nhorman It has been over 24 hours since I got two approvals. Do I need to mark the conversations are resolved to get this to move to "done"?

**Comment by nhorman** [comment]
Yeah, looks like, I'll set the flag and we'll get to the merge soon

File: crypto/x509/standard_exts.h
Comment (FdaSilvaYY) [reply crypto/x509/standard_exts.h:85]: This comma creates an ending <<nul>> item, who is useless.

**Comment by FdaSilvaYY** [comment]
> actually, sorry @JonathanWilbur , I just noticed.  Strictly speaking it has the requisite reviews, but there are some outstanding comments from @FdaSilvaYY that you should probably answer first

I'm OK for a following PR to fix my comments.

File: crypto/x509/standard_exts.h
Comment (t8m) [reply crypto/x509/standard_exts.h:85]: > This comma creates an ending << nul >> item, who is useless.

This is not correct. There is no extra array item created by this trailing comma.

**Comment by t8m** [comment]
Squashed and merged to the master branch. Thank you for your contribution.

**Pull request closed without merging**
